apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,382 @@
+// +kubebuilder:resource:scope="Namespaced"
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type StagedUpdateStrategyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []StagedUpdateStrategy `json:"items"`
+}
+
+// StagedUpdateConfig describes a single update stage group configuration.
+// The clusters in each stage are updated sequentially for now.
+// We will stop the update if any of the updates fail.
+type StagedUpdateConfig struct {
+	// The name of the stage. This MUST be unique within the same StagedUpdateRun.
+	// +required
+	Name string `json:"name"`
+
+	// LabelSelector is a label query over all the joined member clusters. Clusters matching the query are selected.
+	// We don't check if there are overlap between clusters selected by different stages.
+	LabelSelector *metav1.LabelSelector `json:"labelSelector"`
+
+	// The label key used to sort the selected clusters.
+	// The clusters within the stage are updated sequentially following the rule below:
+	//   - primary: Ascending order based on the value of the label key, interpreted as integers if present.
+	//   - secondary: Ascending order based on the name of the cluster if the label key is absent.
+	// +optional
+	SortingLabelKey *string `json:"sortingLabelKey,omitempty"`
+
+	// The collection of tasks that each stage needs to complete successfully before moving to the next stage.
+	// Each task is executed in parallel.
+	// +optional
+	PostStageTasks []PostStageTask `json:"postRolloutTasks,omitempty"`
+}
+
+// PostStageTask is the collection of post stage tasks that ALL need to be completed before we can move to the next stage.
+type PostStageTask struct {
+	// The type of the post stage task.
+	// +kubebuilder:validation:Enum=TimedWait;ManualApproval
+	// +required
+	Type PostStageTaskType `json:"type"`
+
+	// The name of the StagedUpdateApprovalRequest we need to create to get the approval for the next stage.
+	// We will generate a name with a prefix of the StagedUpdateRun name if it's empty.
+	// +optional
+	ApprovalRequestName string `json:"approvalTask,omitempty"`
+
+	// The time to wait after all the clusters in the current stage complete the update before we move to the next stage.
+	// +kubebuilder:default:="1h"
+	// +kubebuilder:validation:Pattern="^0|([0-9]+(\\.[0-9]+)?(s|m|h))+$"
+	// +kubebuilder:validation:Type:=string
+	// +optional
+	WaitTime metav1.Duration `json:"waitTime,omitempty"`
+}
+
+// StagedUpdateRunStatus defines the observed state of the StagedUpdateRun.
+type StagedUpdateRunStatus struct {
+	// ComputedStages list the name of the clusters in each stage when the updateRun started.
+	// The clusters in each stage are ordered following its order to be rolled out.
+	// The updateRun will stop if the cluster stages change during the update.
+	// +required
+	ComputedStages [][]string `json:"currentStages"`
```

Comment:
Use []StageStatus which contains each cluster's status. See this one: https://learn.microsoft.com/en-us/rest/api/fleet/update-runs/get?view=rest-fleet-2023-10-15&tabs=HTTP